overlord/hookstate/ctlcmd/helpers.go

```diff
@@ -122,6 +136,38 @@ func queueCommand(context *hookstate.Context, tts []*state.TaskSet) error {
 		}
 		change.AddAll(ts)
 	}
+
+	// As this can be run from what was originally the last task of a change,
+	// make sure the tasks added to the change are considered immediately.
+	st.EnsureBefore(0)
+
+	return nil
+}
+
+// queueDefaultConfigureHookCommand queues service command exactly after start-snap-services.
+//
+// This is possible because the default-configure hook is run on first-install only and right
+// after start-snap-services is the nearest we can queue the service commands safely to make
+// sure all the needed state is setup properly.
+func queueDefaultConfigureHookCommand(context *hookstate.Context, tts []*state.TaskSet) error {
+	st := context.State()
+	st.Lock()
+	defer st.Unlock()
+
+	_, tasks, err := prepareQueueCommand(context, tts)
+	if err != nil {
+		return err
+	}
+
+	for _, t := range tasks {
+		if t.Kind() == "start-snap-services" {
```

Comment:
I fear that this a bit too simple for the case where the change is using a single lane/transaction=all-snaps, as there might be multiple start-snap-services involved, as there might just one lane for everything. I think this needs to check also that the snap-name of this task matches the snap name of the hook at least,  we probably need a test somewhere for this or a spread test.  configure has probably the same problem already but there is less noticeable because it would end waiting for all the snaps in the change, which maybe we want to fix as well

Reply:
Great catch, Thank you!

I added a test to trigger this edge case and fixed this bug but only for default-configure hook since configure hook is not affected that much and have been like that since the start. The reason I skipped the fix for configure hook is because there is no straightforward way to associate an arbitrary task kind with a snap name.